internal/keepers/filter.go

```diff
@@ -0,0 +1,176 @@
+	c := &reportCoordinator{
+		logger:     logger,
+		registry:   r,
+		logs:       logs,
+		idBlocks:   newCache[bool](s),
+		activeKeys: newCache[bool](time.Hour), // 1 hour allows the cleanup routine to clear stale data
+		chStop:     make(chan struct{}),
+	}
+
+	cl := &intervalCacheCleaner[bool]{
+		Interval: cacheClean,
+		stop:     make(chan struct{}),
+	}
+
+	c.cacheCleaner = cl
+
+	runtime.SetFinalizer(c, func(srv *reportCoordinator) { srv.stop() })
+
+	c.start()
+
+	return c
+}
+
+// Add adds the provided upkeep to the filter.
+func (rc *reportCoordinator) Add(key types.UpkeepKey) error {
+	id, err := rc.registry.IdentifierFromKey(key)
+	if err != nil {
+		return err
+	}
+
+	rc.idBlocks.Set(string(id), true, defaultExpiration)
+
+	return nil
+}
+
+// Filter returns a filter function that removes upkeep keys that apply to this
+// filter
+func (rc *reportCoordinator) Filter() func(types.UpkeepKey) bool {
+	return func(key types.UpkeepKey) bool {
+		id, err := rc.registry.IdentifierFromKey(key)
+		if err != nil {
+			return false
+		}
+
+		if _, ok := rc.idBlocks.Get(string(id)); ok {
+			return false
+		}
+
+		return true
+	}
+}
+
+func (rc *reportCoordinator) Accept(key types.UpkeepKey) {
+	rc.activeKeys.Set(string(key), false, defaultExpiration)
+}
+
+func (rc *reportCoordinator) IsTransmitting(key types.UpkeepKey) bool {
+	// key is assumed to be transmitted if it doesn't exist in cache or if it
+	// is no longer waiting for a transaction to be transmitted
+	inLogs, ok := rc.activeKeys.Get(string(key))
+	return ok && inLogs
+}
+
+func (rc *reportCoordinator) start() {
+	rc.starter.Do(func() {
+		go rc.run()
+		go rc.cacheCleaner.Run(rc.idBlocks)
+		go rc.cacheCleaner.Run(rc.activeKeys)
+	})
+}
+
+func (rc *reportCoordinator) stop() {
+	rc.chStop <- struct{}{}
+	rc.cacheCleaner.stop <- struct{}{}
+	rc.cacheCleaner.stop <- struct{}{}
+}
+
+func (rc *reportCoordinator) run() {
+	// TODO: handle panics by restarting this process
+
+	cadence := time.Second
+	timer := time.NewTimer(cadence)
+
+	for {
+
+		select {
+		case <-timer.C:
+			startTime := time.Now()
+			keys := rc.activeKeys.Keys()
+			logs, _ := rc.logs.PerformLogs(context.Background())
+
+			// log entries indicate that a perform exists on chain in some
+			// capacity. the existance of an entry means that the transaction
+			// was broadcast by at least one node. reorgs can still happen
+			// causing performs to vanish or get moved to a later block.
+			//
+			// in the case of reorgs causing a perform to be dropped from the
+			// chain, total confirmations are checked to allow other nodes
+			// to retry a transaction based on the OCR transmit cadence.
+			logLookup := make(map[string]types.PerformLog)
+			var minConfirmations int64 = 20
+			for _, log := range logs {
+				logLookup[string(log.Key)] = log
+				if _, ok := rc.activeKeys.Get(string(log.Key)); ok {
+					if log.Confirmations < minConfirmations {
+						// set the active key value to false to indicate that
```

Comment:
nit. true/false is opposite as comment and code

Reply:
yup